app/upgrade_test.go

```diff
@@ -0,0 +1,114 @@
+package app_test
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	upgradetypes "github.com/cosmos/cosmos-sdk/x/upgrade/types"
+	"github.com/osmosis-labs/osmosis/app"
+	"github.com/stretchr/testify/suite"
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+)
+
+type UpgradeTestSuite struct {
+	suite.Suite
+
+	ctx sdk.Context
+	app *app.OsmosisApp
+}
+
+func (suite *UpgradeTestSuite) SetupTest() {
+	suite.app = app.Setup(false)
+	suite.ctx = suite.app.BaseApp.NewContext(false, tmproto.Header{Height: 1, ChainID: "osmosis-1", Time: time.Now().UTC()})
+}
+
+func TestKeeperTestSuite(t *testing.T) {
+	suite.Run(t, new(UpgradeTestSuite))
+}
+
+func (suite *UpgradeTestSuite) TestUpgradePayments() {
+	testCases := []struct {
+		msg         string
+		pre_update  func()
+		update      func()
+		post_update func()
+		expPass     bool
+	}{
+		{
+			"Test community pool payouts for Prop 12",
+			func() {
+				// mint coins to distribution module / feepool.communitypool
+
+				var bal = int64(1000000000000)
+				coin := sdk.NewInt64Coin("uosmo", bal)
+				coins := sdk.NewCoins(coin)
+				suite.app.BankKeeper.MintCoins(suite.ctx, "mint", coins)
+				suite.app.BankKeeper.SendCoinsFromModuleToModule(suite.ctx, "mint", "distribution", coins)
+				feePool := suite.app.DistrKeeper.GetFeePool(suite.ctx)
+				feePool.CommunityPool = feePool.CommunityPool.Add(sdk.NewDecCoinFromCoin(coin))
+				suite.app.DistrKeeper.SetFeePool(suite.ctx, feePool)
+
+			},
+			func() {
+				// run upgrade
+
+				plan := upgradetypes.Plan{Name: "v4", Height: 5}
+				suite.app.UpgradeKeeper.ScheduleUpgrade(suite.ctx, plan)
+				plan, exists := suite.app.UpgradeKeeper.GetUpgradePlan(suite.ctx)
+				suite.Require().True(exists)
+				suite.Require().NotPanics(func() {
+					suite.app.UpgradeKeeper.ApplyUpgrade(suite.ctx.WithBlockHeight(5), plan)
+				})
+			},
+			func() {
+				var total = int64(0)
+
+				// check that each account got the payment expected
+				payments := app.GetProp12Payments()
+				for _, payment := range payments {
+					addr, err := sdk.AccAddressFromBech32(payment[0])
+					suite.Require().NoError(err)
+					amount, err := strconv.ParseInt(strings.TrimSpace(payment[1]), 10, 64)
+					suite.Require().NoError(err)
+					coin := sdk.NewInt64Coin("uosmo", amount)
+
+					accBal := suite.app.BankKeeper.GetBalance(suite.ctx, addr, "uosmo")
+					suite.Require().Equal(coin, accBal)
+
+					total += amount
+				}
+
+				//check that the total paid out was as expected
+				suite.Require().Equal(total, int64(367926557424))
```

Comment:
Can we derive the expected total based on daily issuance, and hardcoded ratios?

Reply:
Hmm... yes, tho now that I'm doing that math I'm realizing the number there is off by a few hundred uosmo. (actual amount should be: 367926558129)
The correct calculation is:

(300,000,000,000 * 95000 * 45) / (100  * 365 * 95500)

(300mil osmo per year, 45% to LPs, 95000/95500 of that going to pools with 500/95500 going to community pool)

The constant I currently have there is the sum of the integer uosmo payment amounts output from the script. But it doesn't exactly match the theoretically correct number, because that would require including the sub uosmo amounts each account should be getting in the sum, and rounding to an integer after summing...

So I could change this to use the formula, but then the test will fail because there's... 705 uosmo, that aren't being paid out.

I suppose I could have the script output floats, and then the upgrade code would round them down to ints for the actual payments, but could sum up the floats for checking the total?

Reply:
Hrmm, can we have the python code round up, and then just check here that the amount overpaid is less than 1 osmo?